### PR TITLE
introduce 'called_function' into goto traces [blocks: #3126]

### PR DESCRIPTION
**Pull request by kroening**
This will support removing the function identifier from goto instructions.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — APPROVED** [review]
Looks ok to me except clang-format will have it's complaints.

File: src/goto-programs/goto_trace.cpp [hunk 1]
```diff
@@ -483,7 +485,7 @@ void show_full_goto_trace(
       function_depth--;
       if(options.show_function_calls)
       {
-        out << "\n#### Function return: " << step.function_identifier
+        out << "\n#### Function return"
```
Comment (smowton) [inline src/goto-programs/goto_trace.cpp]: Why not `step.function`?

File: src/goto-programs/goto_trace.h [hunk 2]
```diff
@@ -86,6 +86,8 @@ class goto_trace_stept
   enum class assignment_typet { STATE, ACTUAL_PARAMETER };
   assignment_typet assignment_type;
 
+  // the instruction that created this step
```
Comment (smowton) [inline src/goto-programs/goto_trace.h]: Clarify that calls are outside a function and returns inside it?

File: src/goto-programs/xml_goto_trace.cpp [hunk 3]
```diff
@@ -169,23 +169,42 @@ void convert(
...
+        xml_call_return.set_attribute("thread", std::to_string(step.thread_nr));
+        xml_call_return.set_attribute("step_nr", std::to_string(step.step_nr));
+
+        const symbolt &symbol = ns.lookup(step.called_function);
+        xmlt &xml_function=xml_call_return.new_element("function");
+        xml_function.set_attribute(
+          "display_name", id2string(symbol.display_name()));
+        xml_function.set_attribute(
+          "identifier", id2string(symbol.name));
+        xml_function.new_element()=xml(symbol.location);
```
Comment (smowton) [inline src/goto-programs/xml_goto_trace.cpp]: style (here and below)

File: src/goto-programs/xml_goto_trace.cpp [hunk 4]
```diff
@@ -169,23 +169,42 @@ void convert(
...
         xml_call_return.set_attribute("step_nr", std::to_string(step.step_nr));
 
-        const symbolt &symbol = ns.lookup(step.function_identifier);
+        const symbolt &symbol = ns.lookup(step.function);
         xmlt &xml_function=xml_call_return.new_element("function");
         xml_function.set_attribute(
           "display_name", id2string(symbol.display_name()));
         xml_function.set_attribute(
-          "identifier", id2string(step.function_identifier));
+          "identifier", id2string(symbol.name));
```
Comment (smowton) [inline src/goto-programs/xml_goto_trace.cpp]: `step.function` a little clearer?

**Review by smowton — APPROVED** [review]
A few nitpicks

File: src/goto-programs/goto_trace.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-programs/goto_trace.cpp]: ok

File: src/goto-programs/goto_trace.h (hunk 2 above)
Comment (kroening) [reply src/goto-programs/goto_trace.h]: ok

File: src/goto-programs/xml_goto_trace.cpp (hunk 4 above)
Comment (kroening) [reply src/goto-programs/xml_goto_trace.cpp]: ok

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 05eb134c3817b6485d9cf64fb1670f33e24ce1a7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90559178
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**